### PR TITLE
Expose document highlighting to server.

### PR DESCRIPTION
**Pull request by hoanhtien**
Will update this pull request with more tests.

### CODE REVIEW

File: src/harness/fourslash.ts [hunk 1]
```diff
@@ -2132,7 +2132,7 @@ module FourSlash {
                 let occurance = occurances[i];
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: occurrance -> occurrence

File: src/harness/fourslash.ts [hunk 2]
```diff
@@ -2132,7 +2132,7 @@ module FourSlash {
                 let occurance = occurances[i];
                 if (occurance && occurance.fileName === fileName && occurance.textSpan.start === start && ts.textSpanEnd(occurance.textSpan) === end) {
                     if (typeof isWriteAccess !== "undefined" && occurance.isWriteAccess !== isWriteAccess) {
-                        this.raiseError('verifyOccurancesAtPositionListContains failed - item isWriteAccess value doe not match, actual: ' + occurance.isWriteAccess + ', expected: ' + isWriteAccess + '.');
+                        this.raiseError('verifyOccurancesAtPositionListContains failed - item isWriteAccess value does not match, actual: ' + occurance.isWriteAccess + ', expected: ' + isWriteAccess + '.');
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: verifyOccurrencesAtPositionListContains

File: src/harness/fourslash.ts [hunk 3]
```diff
@@ -2152,6 +2152,51 @@ module FourSlash {
...
+        public verifyDocumentHighlightsAtPositionListContains(fileName: string, start: number, end: number, fileNamesToSearch: string[], kind?: string) {
+            this.taoInvalidReason = 'verifyDocumentHighlightsAtPositionListContains NYI';
+
+            let documentHighlights = this.getDocumentHighlightsAtCurrentPosition(fileNamesToSearch);
+
+            if (!documentHighlights || documentHighlights.length === 0) {
+                this.raiseError('verifyDocumentHighlightsAtPositionListContains failed - found 0 highlights, expected at least one.');
+            }
+
+            for (let i = 0; i < documentHighlights.length; i++) if (documentHighlights[i].fileName === fileName) {
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Use a for..of loop, `if` on a separate line

File: src/harness/fourslash.ts [hunk 4]
```diff
@@ -2152,6 +2152,51 @@ module FourSlash {
...
+            this.raiseError('verifyOccurancesAtPositionListContains failed - could not find the item: ' + JSON.stringify(missingItem) + ' in the returned list: (' + JSON.stringify(documentHighlights) + ')');
+        }
+
+        public verifyDocumentHighlightsAtPositionListCount(expectedCount: number, fileNamesToSearch: string[]) {
+            this.taoInvalidReason = 'verifyDocumentHighlightsAtPositionListCount NYI';
+
+            let documentHighlights = this.getDocumentHighlightsAtCurrentPosition(fileNamesToSearch);
+            let actualCount = documentHighlights 
+                ? documentHighlights.reduce((currentCount, currentDocumentHighlights) => { 
+                    return currentCount + currentDocumentHighlights.highlightSpans.length}, 0)
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: If you destructure `currentDocumentHighlights`, this might be easier.

File: src/server/client.ts [hunk 5]
```diff
@@ -528,7 +528,29 @@ namespace ts.server {
...
+            var args: protocol.FileLocationRequestArgs = {
+                file: fileName,
+                line: lineOffset.line,
+                offset: lineOffset.offset,
+            };
+
+            var request = this.processRequest<protocol.DocumentHighlightsRequest>(CommandNames.DocumentHighlights, args);
+            var response = this.processResponse<protocol.DocumentHighlightsResponse>(request);
+
+            return response.body.map(entry => { // convert ts.server.protocol.DocumentHighlightsItem to ts.DocumentHighlights
```
Comment (DanielRosenwasser) [inline src/server/client.ts]: Put the comment on the preceding line

File: src/server/client.ts [hunk 6]
```diff
@@ -528,7 +528,29 @@ namespace ts.server {
...
+                return {
+                    fileName: entry.file,
+                    highlightSpans: entry.highlightSpans.map(span => { // convert ts.server.protocol.HighlightSpan to ts.HighlighSpan
```
Comment (DanielRosenwasser) [inline src/server/client.ts]: Comment on preceding line

File: src/server/protocol.d.ts [hunk 7]
```diff
@@ -239,6 +239,28 @@ declare namespace ts.server.protocol {
...
 
     /**
+      * Get document highlights request; value of command field is
+      * "documentHighlights". Return response giving spans that are relevant
+      * in the file at a given line and column.
+      */
+    export interface DocumentHighlightsRequest extends FileLocationRequest {
+    }
+
+    export interface HighLightSpan extends TextSpan {
```
Comment (DanielRosenwasser) [inline src/server/protocol.d.ts]: HighlightSpan

File: src/server/session.ts [hunk 8]
```diff
@@ -343,6 +344,42 @@ namespace ts.server {
...
+            let documentHighlights = compilerService.languageService.getDocumentHighlights(fileName, position, filesToSearch);
+            
+            if (!documentHighlights) {
+                return undefined;
+            }
+
+            return documentHighlights.map(documentHighlight => { // convert ts.DocumentHighlights to ts.server.protocol.DocumentHighlightsItem
+                var file = documentHighlight.fileName;
+                return {
+                    file: file,
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: Use a shorthand property

File: src/server/session.ts [hunk 9]
```diff
@@ -343,6 +344,42 @@ namespace ts.server {
...
+                    highlightSpans: documentHighlight.highlightSpans.map(highlightSpan => { // convert to ts.HighlightSpan to ts.server.protocol.HighlightSpan
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: Consider factoring this our into a local variable named` highlightSpans` with a function `convertToBlah`

Comment (DanielRosenwasser) [inline src/server/session.ts]: Comment above code

File: src/server/session.ts [hunk 10]
```diff
@@ -343,6 +344,42 @@ namespace ts.server {
...
+                        let { textSpan, kind } = highlightSpan;
+                        let start = compilerService.host.positionToLineOffset(file, textSpan.start);
+                        let end = compilerService.host.positionToLineOffset(file, ts.textSpanEnd(textSpan));
+                        return {
+                            start: start,
+                            end: end,
+                            kind: kind
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: Use shorthand properties

File: src/server/session.ts [hunk 11]
```diff
@@ -343,6 +344,42 @@ namespace ts.server {
...
+            let filesToSearch = [fileName]; // only search for highlights inside the current file
+            
+            let documentHighlights = compilerService.languageService.getDocumentHighlights(fileName, position, filesToSearch);
+            
+            if (!documentHighlights) {
+                return undefined;
+            }
+
+            return documentHighlights.map(documentHighlight => { // convert ts.DocumentHighlights to ts.server.protocol.DocumentHighlightsItem
+                var file = documentHighlight.fileName;
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: Use `

File: src/server/client.ts [hunk 12]
```diff
@@ -528,7 +528,29 @@ namespace ts.server {
         }
 
         getDocumentHighlights(fileName: string, position: number): DocumentHighlights[] {
-            throw new Error("Not Implemented Yet.");
+            var lineOffset = this.positionToOneBasedLineOffset(fileName, position);
```
Comment (DanielRosenwasser) [inline src/server/client.ts]: Use `let` not `var`

Comment (DanielRosenwasser) [inline src/server/client.ts]: If you destructured this into `line` and `offset` you could use shorthand properties below

File: src/server/client.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/server/client.ts]: The comment sounds backwards

File: tests/cases/fourslash/fourslash.ts [hunk 13]
```diff
@@ -421,6 +421,14 @@ module FourSlashInterface {
             FourSlash.currentTestState.verifyOccurrencesAtPositionListCount(expectedCount);
         }
 
+        public documentHighlightsAtPositionContains(range: Range, fileNamesToSearch: string[], kind?: string) {
+            FourSlash.currentTestState.verifyDocumentHighlightsAtPositionListContains(range.fileName, range.start, range.end, fileNamesToSearch, kind);
+        }
+
+        public documentHighlightsAtPositionCount(expectedCount: number, fileNamesToSearch: string[], kind?: string) {
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/fourslash.ts]: You aren't using `kind` here

**Comment by mhegazy** [comment]
other than @DanielRosenwasser comments, looks good to me.

**Comment by hoanhtien** [comment]
Addressed comments.

File: src/harness/fourslash.ts [hunk 14]
```diff
@@ -2122,36 +2122,83 @@ module FourSlash {
...
+            if (!occurrences || occurrences.length === 0) {
                 this.raiseError('verifyOccurancesAtPositionListContains failed - found 0 references, expected at least one.');
             }
 
-            for (let i = 0; i < occurances.length; i++) {
-                let occurance = occurances[i];
-                if (occurance && occurance.fileName === fileName && occurance.textSpan.start === start && ts.textSpanEnd(occurance.textSpan) === end) {
-                    if (typeof isWriteAccess !== "undefined" && occurance.isWriteAccess !== isWriteAccess) {
-                        this.raiseError('verifyOccurancesAtPositionListContains failed - item isWriteAccess value doe not match, actual: ' + occurance.isWriteAccess + ', expected: ' + isWriteAccess + '.');
+            for (let i = 0; i < occurrences.length; i++) {
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Use a for-of here

File: src/harness/fourslash.ts [hunk 15]
```diff
@@ -2122,36 +2122,83 @@ module FourSlash {
...
-            let actualCount = occurances ? occurances.length : 0;
+            let occurrences = this.getOccurancesAtCurrentPosition();
+            let actualCount = occurrences ? occurrences.length : 0;
             if (expectedCount !== actualCount) {
                 this.raiseError('verifyOccurrencesAtPositionListCount failed - actual: ' + actualCount + ', expected:' + expectedCount);
             }
         }
 
+        private getDocumentHighlightsAtCurrentPosition(fileNamesToSearch: string[]) {
+            let filesToSearch = fileNamesToSearch.map(name => this.basePath + "/" + name);
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: `combinePaths` might be more appropriate, check with @mhegazy

File: src/harness/fourslash.ts [hunk 16]
```diff
@@ -2122,36 +2122,83 @@ module FourSlash {
...
+            let missingItem = { fileName: fileName, start: start, end: end, kind: kind };
+            this.raiseError('verifyOccurancesAtPositionListContains failed - could not find the item: ' + JSON.stringify(missingItem) + ' in the returned list: (' + JSON.stringify(documentHighlights) + ')');
+        }
+
+        public verifyDocumentHighlightsAtPositionListCount(expectedCount: number, fileNamesToSearch: string[]) {
+            this.taoInvalidReason = 'verifyDocumentHighlightsAtPositionListCount NYI';
+
+            let documentHighlights = this.getDocumentHighlightsAtCurrentPosition(fileNamesToSearch);
+            let actualCount = documentHighlights
+                ? documentHighlights.reduce((currentCount, { fileName, highlightSpans }) => {
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Don't use a block body

File: src/harness/fourslash.ts [hunk 17]
```diff
@@ -2122,36 +2122,83 @@ module FourSlash {
...
+
+            let missingItem = { fileName: fileName, start: start, end: end, kind: kind };
+            this.raiseError('verifyOccurancesAtPositionListContains failed - could not find the item: ' + JSON.stringify(missingItem) + ' in the returned list: (' + JSON.stringify(documentHighlights) + ')');
+        }
+
+        public verifyDocumentHighlightsAtPositionListCount(expectedCount: number, fileNamesToSearch: string[]) {
+            this.taoInvalidReason = 'verifyDocumentHighlightsAtPositionListCount NYI';
+
+            let documentHighlights = this.getDocumentHighlightsAtCurrentPosition(fileNamesToSearch);
+            let actualCount = documentHighlights
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Invert the condition to make this more readable

File: src/server/session.ts [hunk 18]
```diff
@@ -343,6 +344,43 @@ namespace ts.server {
...
+
+            function convertToDocumentHighlightsItem(documentHighlights: ts.DocumentHighlights): ts.server.protocol.DocumentHighlightsItem {
+                let { fileName, highlightSpans } = documentHighlights;
+
+                return {
+                    file: fileName,
+                    highlightSpans: highlightSpans.map(convertHighlightSpan1)
+                };
+
+                function convertHighlightSpan1(highlightSpan: ts.HighlightSpan): ts.server.protocol.HighlightSpan {
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: Why did you need to differentiated with numbers?

File: src/server/client.ts [hunk 19]
```diff
@@ -528,7 +528,32 @@ namespace ts.server {
...
 
         getDocumentHighlights(fileName: string, position: number): DocumentHighlights[] {
-            throw new Error("Not Implemented Yet.");
+            let { line, offset } = this.positionToOneBasedLineOffset(fileName, position);
+            let args: protocol.FileLocationRequestArgs = { file: fileName, line, offset };
+
+            let request = this.processRequest<protocol.DocumentHighlightsRequest>(CommandNames.DocumentHighlights, args);
+            let response = this.processResponse<protocol.DocumentHighlightsResponse>(request);
+
+            let _self = this;
```
Comment (DanielRosenwasser) [inline src/server/client.ts]: If you have to capture `this` it might not be worth it. I'd convert it back to an arrow function.

Also, why `_self` instead of `self`?

Comment (hoanhtien) [inline src/server/client.ts]: I think arrow function is not very readable. Is there any downside with capturing `this`?

Comment (DanielRosenwasser) [inline src/server/client.ts]: It's just slightly annoying. I'd still change to `self` though if that doesn't conflict with anything

File: src/harness/fourslash.ts (hunk 15 above)
Comment (mhegazy) [inline src/harness/fourslash.ts]: +1, though we do not use it consistently elsewhere.

**Comment by mhegazy** [comment]
:+1:
